File: 51. N-Queens.py
class Solution(object):
    def solveNQueens(self, n):
        """
        :type n: int
        :rtype: List[List[str]]
        """
        self.res = []
        self.DFS([], n)
        return self.res
        
    def check(self, path, i):
        for j in xrange(len(path)):
            if i == path[j] or len(path)-j == abs(path[j] - i):
                return False
        return True
    def DFS(self, path, n):
        if len(path) == n:
            temp=[]
            for i in xrange(len(path)):
                row=""
                for j in xrange(n):
                   if j!=path[i]:
                       row+="."
                   else:
                       row+="Q"
                temp+=[row]
            self.res+=[temp[:]]
        for i in xrange(n):
            if self.check(path, i):
                self.DFS(path + [i], n)

File: 51. N-Queens.cpp
class Solution {
private:
    vector<vector<string>> res;
public:
    vector<vector<string>> solveNQueens(int n) {
        DFS({}, n);
        return res;
    }
    bool check(vector <int> path, int i)
    {
        for (int j = 0; j < path.size(); j++)
        {
            if (path[j] == i || path.size() - j == abs(path[j] - i))
                return false;
        }
        return true;
    }
    void DFS(vector <int> path, int& n)
    {
        if (path.size() == n)
        {
            vector <string> board;
            for (int i = 0; i < path.size(); i++)
            {
                string row = "";
                for (int j = 0; j < n; j++)
                {
                    if (j != path[i])
                        row += '.';
                    else
                        row += 'Q';
                }
                board.push_back(row);
            }
            res.push_back(board);
            return;
        }
        for (int i = 0; i < n; i++)
        {
            if (check(path, i))
            {
                path.push_back(i);
                DFS(path, n);
                path.pop_back();
            }
        }
    }
};
